### ISSUE TITLE
SPFx webparts not able to acquire token for Graph API in Teams app (Viva Connections)

### ISSUE THREAD

**Issue by Vighough** [SharePoint/sp-dev-docs#9099]
### What type of issue is this?

other

### What SharePoint development model, framework, SDK or API is this about?

💥 SharePoint Framework

### Target SharePoint environment

SharePoint Online

### What browser(s) / client(s) have you tested

- [ ] 💥 Internet Explorer
- [X] 💥 Microsoft Edge
- [ ] 💥 Google Chrome
- [ ] 💥 FireFox
- [ ] 💥 Safari
- [ ] mobile (iOS/iPadOS)
- [ ] mobile (Android)
- [ ] not applicable
- [ ] other (enter in the "Additional environment details" area below)

### Additional environment details

- Microsoft Egde version 115.0.1901.183 (Official build) (64-bit)
- SPFx 1.17.4
- Node 16.15.1
- NPM 8.11.0

### Issue description

We have created an SharePoint intranet site and made it the root site. We then added the site to the Viva Connections integration in Teams, as an app. On this site we have a couple of custom webparts that we created. One of the webparts have been used in many intranet sites and haven't had these problems so far. The problem also isn't occurring on SharePoint itself.

The problem is connecting to the Graph API, the token doesn't seem to be retrieved and thus the API requests are not even send to the Graph API.

The error I'm getting:

`Uncaught (in promise) Error: Token request previously failed
    at new t (chunk.aadTokenProvider_en-us_445a72773ba77a61be93.js:1:1514)
    at e._getTokenInternal (chunk.aadTokenProvider_en-us_445a72773ba77a61be93.js:1:3567)
    at e.getToken (chunk.aadTokenProvider_en-us_445a72773ba77a61be93.js:1:2372)
    at sp-pages-assembly_en-us_c3c628b67d521769ee27f473c0ed9543.js:76:647657`

The request that might be the problem:

`https://tenant.sharepoint.com/_api/Microsoft.SharePoint.Internal.ClientSideComponent.Token.AcquireOBOToken?resource=%27https://graph.microsoft.com%27&clientId=%2775aff87b-45bf-41e7-84f3-552901a72ea0%27`

Results in an 500 internal server error with the following message:

`{
    "odata.error": {
        "code": "-1, System.AggregateException",
        "message": {
            "lang": "en-US",
            "value": "One or more errors occurred."
        }
    }
}`

I found this issue #7884 (https://github.com/SharePoint/sp-dev-docs/issues/7884) which seems to be related to mine or at least gives the same error, but the resolution there doesn't apply here sadly. The manifest is already in the right format and with the correct data. Also cross referenced the manifest with other, working tenants

At this point I have no clue on what the cause of the issue is, could any of you help me with this problem?

**Comment by heesungjang** [comment]
Our team is facing the same issue that @Vighough encountered. Any assistance would be greatly appreciated.

**Comment by lucabandMSFT** [comment]
@Vighough , thanks for reaching out.

Can you please explain to me what this means? "We then added the site to the Viva Connections integration in Teams, as an app"?

**Comment by Vighough** [comment]
Hi @lucabandMSFT, 

We are using the Viva Connections integration of the SharePoint intranet site. This allows organisations to have the SharePoint intranet site as an app in Teams, for better usability. You can try this integration by opening a SharePoint site in the Viva Connections app in Teams see (https://support.microsoft.com/en-us/office/access-and-use-the-viva-connections-app-in-microsoft-teams-8b4e7f76-f305-49a9-b6d2-09378476f95b)

See: [Set up Viva Connections in the Microsoft 365 admin center](https://learn.microsoft.com/en-us/viva/connections/set-up-admin-center)

**Comment by lucabandMSFT** [comment]
@Vighough, thanks. This is what confuses me.

Viva Connections (in the link you provided) does not support or even allow to visualize pages with Web Parts. The Viva Connection experience relies on a fixed page (dashboard) that can be extended with specific widget called cards or Adaptive Card Extensions which are specific type of SPFx components.

So I'm wondering where you are seeing the error. Mind provide a description of the flow or a link to a video that shows the behavior please?

thanks.

**Comment by Vighough** [comment]
Hi @lucabandMSFT, 

You are right and to clarify we are not using that dashboard and those widgets (although it is available to the users, it's not to our concern)

We are using the option to add the SharePoint home site to the Viva Connections app. [See this specific chapter](https://learn.microsoft.com/en-us/viva/connections/set-up-admin-center#add-a-sharepoint-home-site-after-setting-up-a-connections-experience)

I'm seeing this error inside the Viva Connections app, inside Teams, while opening the SharePoint home site or any other site in the Viva Connections app. I cannot share a video of the production environment sadly, but I'll be happy to show you the issue during a short call?

Also, here is a screenshots of how to reach a SharePoint site inside the Viva Connections app. Open any SharePoint site that has custom SPFx webparts that use the Graph API. It is worth mentioning that I cannot reproduce this on other tenants, as I'm not able to find the root cause, so this might work fine in your case. Nevertheless, this should give you some context

<img width="729" alt="image" src="https://github.com/SharePoint/sp-dev-docs/assets/11834321/97a1f14c-f143-4d01-b1b7-3ebbd826cba5">

Many thanks in advance

**Comment by lucabandMSFT** [comment]
thanks @Vighough,

One last question, would be possible for you to provide a snipped of the SPFx custom code that fails when the page runs on Teams?

**Comment by c-eiser13** [comment]
@lucabandMSFT I'm having the same issue, when clicking the link in the upper right of viva connections to go to the home site, I have an SPFx web part on the home site that calls MS Graph, and this web part generates an error attempting to get the token. Viva Connections appears to be an iframe, perhaps that is part of the token acquisition issue? Here is the error I see in the Teams desktop client dev tools

![image](https://github.com/SharePoint/sp-dev-docs/assets/42096464/a2b517e6-39d6-4845-85fc-8d7561ff643a)

Here is my code that is making the graph call:

![image](https://github.com/SharePoint/sp-dev-docs/assets/42096464/252930f2-3c9e-4ae2-9507-9d795a36f1dd)

If there is any further info I can provide to help troubleshoot, please let me know. Thanks!

**Comment by Vighough** [comment]
Hi @lucabandMSFT,

I have this problem with several custom made webparts. To try to isolate the problem, I already tested with a minimal webpart and with this minimal implementation the issue is also occurring. I have made a [repository](https://github.com/Vighough/minimal-sample-webpart) of this webpart, so you can see the full code

**Comment by Vighough** [comment]
For anyone else having this problem: 

We are using this as a temporary fix [Add SharePoint site as a App to the Teams - Microsoft Community Hub](https://techcommunity.microsoft.com/t5/microsoft-teams/add-sharepoint-site-as-a-app-to-the-teams/m-p/2864482)

It has the same SharePoint intranet stie integration effect as the Viva Connections app, just not the other features that Viva provides

**Comment by GrahamMcMynn** [comment]
@Vighough - Would it be possible to send a network trace of the call to https://tenant.sharepoint.com/_api/Microsoft.SharePoint.Internal.ClientSideComponent.Token.AcquireOBOToken that is failing?

**Comment by Vighough** [comment]
Hi @GrahamMcMynn,

Thank you for the detailed steps. I've discussed this with the client and we're planning a moment with an admin to try this out. I'll let know if this has worked or not!

The tenant is an Education version and it's pretty old as what I've heard. They have had more problems that might be related to that. We have added some permissions manually to that app registration, because there is some sort of bug with the API access approval from SharePoint. See my posts on these two forums about this:

[MS Forum](https://answers.microsoft.com/en-us/msoffice/forum/all/approved-sharepoint-api-access-not-correctly-added/680b77f9-6a2d-4730-820b-f8694fedb889)
[MS Feedback](https://feedbackportal.microsoft.com/feedback/idea/b4e0967f-1667-ed11-a81b-000d3a7dd78c)
